### ISSUE TITLE
Failure to set an option from API is sometimes translated but sometimes not

### ISSUE THREAD

**Issue by zeertzjq** [neovim/neovim#18128]
### Neovim version (nvim -v)

v0.8.0-dev+2-g08a144912c

### Vim (not Nvim) behaves the same?

N/A

### Operating system/version

Arch Linux

### Terminal name/version

kitty 0.25.0

### $TERM environment variable

xterm-kitty

### Installation

AUR neovim-git

### How to reproduce the issue

1. Run `nvim --clean` with a locale
2. Run `:call nvim_set_option('statusline', '%$')`
3. The `E539` error message is translated
4. Run `:call nvim_set_option('statusline', '%{')`
5. The `E540` error message is not translated

### Expected behavior

Failure to set an option from API is either always translated or never translated

### Actual behavior

Failure to set an option from API is sometimes translated but sometimes not

**Comment by justinmk** [comment]
should be never translated. API errors (and messages) should be stable.

Note that "stable error messages" is currently not part of `:help api-contract` because we haven't built a good mechanism for standard error messages (i.e. a table of messages and validation functions). Once we do that, it becomes part of the contract.

**Comment by zeertzjq** [comment]
Is this considered an API error? Vim may also change the error message for failure to set an option.

**Comment by justinmk** [comment]
True. And translation doesn't matter much for errors with the `EXX` id. I'd lean towards never translating if possible, just to leave the door open. But as you noted, it's more important to just have consistent behavior.

**Comment by zeertzjq** [comment]
Some translations are done in deeply nested calls:
```
static char *illegal_char(char *errbuf, size_t errbuflen, int c)
{
  if (errbuf == NULL) {
    return "";
  }
  vim_snprintf(errbuf, errbuflen, _("E539: Illegal character <%s>"),
               (char *)transchar(c));
  return errbuf;
}
```
so preventing translation may make the code more ugly.

**Comment by justinmk** [comment]
> so preventing translation may make the code more ugly.

ok, let's do whatever results in simpler code.

I guess I was thinking we could modify `_()` to check if the current context is an RPC/API function, and skip translation.

Translating actually has a non-trivial performance cost, so there's that.

**Comment by zeertzjq** [comment]
This issue also applies to `nvim_eval_statusline()` if it needs to report an error (#18112).
